stdlib/universe/narrow_state_tracking.go

```diff
@@ -0,0 +1,237 @@
+package universe
+
+import (
+	"context"
+
+	"github.com/apache/arrow/go/arrow/memory"
+	"github.com/influxdata/flux"
+	"github.com/influxdata/flux/array"
+	"github.com/influxdata/flux/arrow"
+	"github.com/influxdata/flux/codes"
+	"github.com/influxdata/flux/compiler"
+	"github.com/influxdata/flux/execute"
+	"github.com/influxdata/flux/execute/table"
+	"github.com/influxdata/flux/internal/errors"
+	"github.com/influxdata/flux/values"
+)
+
+func NewNarrowStateTrackingTransformation(ctx context.Context, spec *StateTrackingProcedureSpec, id execute.DatasetID, mem memory.Allocator) (execute.Transformation, execute.Dataset, error) {
+	fn := execute.NewRowPredicateFn(spec.Fn.Fn, compiler.ToScope(spec.Fn.Scope))
+	a := &narrowStateTrackingTransformationAdapter{
+		ctx:      ctx,
+		fn:       fn,
+		timeCol:  spec.TimeCol,
+		countCol: spec.CountColumn,
+		durCol:   spec.DurationColumn,
+		unit:     int64(spec.DurationUnit.Duration()),
+	}
+	nt, d, err := execute.NewNarrowStateTransformation(id, a, mem)
+	if err != nil {
+		return nil, nil, err
+	}
+	return nt, d, nil
+}
+
+type narrowStateTrackingTransformationAdapter struct {
+	ctx context.Context
+	fn  *execute.RowPredicateFn
+
+	timeCol,
+	countCol,
+	durCol string
+
+	unit int64
+}
+
+type trackedState struct {
+	start,
+	prevTime values.Time
+
+	countInState,
+	durationInState bool
+
+	count,
+	duration int64
+}
+
+func (a *narrowStateTrackingTransformationAdapter) Process(chunk table.Chunk, state interface{}, d *execute.TransportDataset, mem memory.Allocator) (interface{}, bool, error) {
+	mod := false // Track whether or not the state has been modified
+	if chunk.Len() == 0 {
+		chunk.Retain()
+		d.Process(chunk)
+		return state, mod, nil
+	}
+
+	if state == nil {
+		state = trackedState{
+			count:    -1,
+			duration: -1,
+		}
+		mod = true
+	}
+	s := state.(trackedState)
+
+	var (
+		times     array.Int
+		fn        execute.RowPredicatePreparedFn
+		counts    array.IntBuilder
+		durations array.IntBuilder
+	)
+	if err := a.scanChunk(chunk, mem, &times, &fn, &counts, &durations); err != nil {
+		chunk.Retain()
+		d.Process(chunk)
+		return s, mod, err
+	}
+
+	buf := chunk.Buffer()
+	for i := 0; i < chunk.Len(); i++ {
+		match, err := fn.EvalRow(a.ctx, i, &buf)
+		if err != nil {
+			chunk.Retain()
+			d.Process(chunk)
+			return s, mod, err
+		}
+		ts := values.Time(times.Value(i))
+		if mod, err := a.updateState(&s, match, ts, &counts, &durations); err != nil {
+			chunk.Retain()
+			d.Process(chunk)
+			return s, mod, err
+		}
+	}
+
+	chunk.Retain()
```

Comment:
I suspect this is to retain each of the arrays within the chunk, but it comes across as a different meaning. Can you retain the individual arrays at the point where you copy them instead of retaining it here?